packages/lib-classifier/src/store/AnnotatedSteps/AnnotatedSteps.js

```diff
@@ -26,46 +40,87 @@ const AnnotatedSteps = types.model('AnnotatedSteps', {
 }))
 .actions(self => {
   setUndoManager(self)
-
+  /** Undo the current step and select the previous step. */
   function back() {
     const { history } = self
+    const { workflowSteps } = getRoot(self)
     if (history.canUndo) {
       history.undo()
+      const { step } = self.latest
+      workflowSteps.selectStep(step.stepKey)
     }
   }
-
-  function createStep({ step, annotations }) {
-    const { steps } = self
-    const historyStep = {
-      id: cuid(),
-      step,
-      annotations
+  /** Create a new history entry from the current active step. **/
+  function createStep() {
+    const { classifications, workflowSteps } = getRoot(self)
+    const classification = tryReference(() => classifications?.active)
+    const step = tryReference(() => workflowSteps.active)
+    if (classification && step) {
+      const { history, steps } = self
+      let annotations
+      history.withoutUndo(() => {
+        annotations = step.tasks.map(task => classification.createAnnotation(task))
+      })
+      const historyStep = {
+        id: cuid(),
+        step,
+        annotations
+      }
+      console.log(`Adding ${historyStep.step.stepKey} to history`)
+      steps.put(historyStep)
     }
-    steps.put(historyStep)
   }
-
-  function next({ step, annotations }) {
+  /** Redo the next step, or add a new step to history if there is no redo. */
+  function next() {
     const { history } = self
+    const expectedNext = self.latest.step.next
+    console.log({ expectedNext })
     if (history.canRedo) {
       history.redo()
+      console.log(`redo ${self.latest.step.stepKey}`)
+      if (expectedNext !== self.latest.step.stepKey) {
+        console.log(`replace ${self.latest.step.stepKey} with ${expectedNext}`)
+        self.replace()
+      }
     } else {
-      createStep({ step, annotations })
+      const { workflowSteps } = getRoot(self)
+      workflowSteps.selectStep(self.latest.step.next)
+      self.createStep()
     }
   }
-
+  /** Clear the redo history and restart history from this point with a new step. */
+  function replace() {
+    const { history } = self
+    const { workflowSteps } = getRoot(self)
+    history.undo()
+    history.clearRedo()
+    workflowSteps.selectStep(self.latest.step.next)
+    self.createStep()
+  }
+  /** Clear stored steps and history. Should be run before classifying a new subject. */
   function reset() {
     const { history, steps } = self
     history.withoutUndo(() => {
       steps.clear()
```

Comment:
`reset` and `start` are called from disposers, which can lead to the history changing unexpectedly when observed values change in other stores. I've stopped that from happening by wrapping those actions in `withoutUndo`. I'll add a comment about that to the store. It might also be possible to make these actions private, if they aren't expected to be called from React components or other models.

Reply:
Done.